### PR TITLE
Remove deprecated callback hooks

### LINKED ISSUE #14841: 🚀 Remove deprecated APIs for 1.8 release
Let's remove the deprecated [1.8 APIs](https://github.com/Lightning-AI/lightning/blob/master/tests/tests_pytorch/deprecated_api/test_remove_1-8.py) together. This is a single place where we can easily track the status of the API.

Please, comment on this issue in order to reserve a particular file to work on. Once you do so, I will edit this top comment to avoid collisions. Once you think your work is finished, please open a PR referencing this issue.

Please note, that it can happen that you may need to edit more than just one file. This is fine, but please keep in mind, that the goal of your PR will be to make the check pass for the chosen file. 

The steps to remove deprecated APIs below
- Decide on the [test](https://github.com/Lightning-AI/lightning/blob/master/tests/tests_pytorch/deprecated_api/test_remove_1-8.py) you want to remove
- grep for all the occurrences of deprecation in the source code and remove
- update [changelog](https://github.com/Lightning-AI/lightning/blob/master/src/pytorch_lightning/CHANGELOG.md)
- make sure that unit tests pass

Example - #14833

For those of you who are not familiar with the process of contributing a PR, we have prepared a [simple guide](https://devblog.pytorchlightning.ai/quick-contribution-guide-86d977171b3a) that will walk you through the necessary steps. You can do it! 🚀 💪

- [x] test_v1_8_0_on_init_start_end #14867
- [x] test_v1_8_0_deprecated_call_hook #14869
- [x] test_v1_8_0_deprecated_run_stage #14870
- [x] test_v1_8_0_trainer_verbose_evaluate #14884
- [x] test_v1_8_0_trainer_ckpt_path_attributes #14897
- [x] test_v1_8_0_deprecated_trainer_should_rank_save_checkpoint #14885
- [x] test_v1_8_0_trainer_optimizers_mixin #14887
- [x] test_v1_8_0_deprecate_trainer_data_loading_mixin #14888
- [x] test_v_1_8_0_deprecated_device_stats_monitor_prefix_metric_keys #14890
- [x] test_v1_8_0_deprecated_lightning_optimizers #14889
- [x] test_v1_8_0_remove_on_batch_start_end #14834
- [x] test_v1_8_0_on_configure_sharded_model #14834
- [x] test_v1_8_0_remove_on_epoch_start_end_lightning_module #14834
- [x] test_v1_8_0_remove_on_pretrain_routine_start_end_lightning_module #14834
- [x] test_v1_8_0_on_before_accelerator_backend_setup #14834
- [x] test_v1_8_0_logger_agg_parameters #14840
- [x] test_v1_8_0_deprecated_agg_and_log_metrics_override #14840
- [x] test_v1_8_0_callback_on_pretrain_routine_start_end #14834
- [x] test_simple_profiler_iterable_durations #14864
- [x] test_v1_8_0_datamodule_checkpointhooks #14909
- [x] test_v1_8_0_deprecated_lightning_ipu_module #14830
- [ ] test_v1_8_0_callback_on_load_checkpoint_hook  #14835
- [ ] test_v1_8_0_callback_on_save_checkpoint_hook #14835

This issue is inspired by the community effort in #13445

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

Removes deprecated API.
Converted all deprecation warnings to errors. 

Per recent decision, we will extend our deprecation window in the following way:
- Deprecate API in v1.x with a warning for removal in v1.(x+2)
- Remove API in v1.(x+2) and convert warnings to errors
- Remove errors for removed APIs in v2.0

Closes #14841 
Part of #15049 

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
I made sure I had fun coding 🙃

cc @borda @awaelchli @ananthsub @rohitgr7 @justusschock @carmocca @ninginthecloud

### CODE REVIEW

File: examples/app_components/python/component_tracer.py [hunk 1]
```diff
@@ -27,6 +27,7 @@ def on_train_start(self, trainer, pl_module) -> None:
                 print("Even the Lightning Work is available and state transfer works !")
                 print(self.lightning_work)
 
+            # FIXME
```
Comment (carmocca) [inline examples/app_components/python/component_tracer.py:30]: FIXME. This app example was using the deprecated hook

**Review by carmocca — APPROVED** [review]
Finished up the removal. Didn't run tests though

**Review by manskx — APPROVED** [review]

**Review by kaushikb11 — APPROVED** [review]

**Review by otaj — APPROVED** [review]

**Review by justusschock — APPROVED** [review]

**Comment by otaj** [comment]
@carmocca, sorry, don't know what happened.

**Review by akihironitta — APPROVED** [review]

**Review by tchaton — APPROVED** [review]
LGTM !

**Comment by awaelchli** [comment]
I think this is blocked by #14947

**Review by lantiga — COMMENTED** [review]
We can think about a migration tool or an error message, as opposed to just failing? A lot of users are still on pre-1.6 and they will simply fail to upgrade.

One strategy could be to convert deprecation warnings to deprecation errors (in `_check_deprecated_callback_hooks`) so that we always have instructions on the why and the how to resolve the errors.

Actually we might want to consider getting those messages back from prior breaking changes.

**Comment by awaelchli** [comment]
@lantiga I created this issue that summarizes the new plan we discussed offline #15049. PR #14835 would be the first to implement this pattern and then we follow with this one here. Further PRs will then bring back partially what was removed here: #14841

**Pull request merged**